raiden-ts/src/utils/error.ts

```diff
@@ -0,0 +1,26 @@
+export type ErrorDetail = { [key: string]: string };
+
+export default abstract class RaidenError extends Error {
+  code: string;
+  details?: ErrorDetail[];
+
+  constructor(message: string, detail?: ErrorDetail[] | ErrorDetail) {
```

Comment:
Suggestion: this `ErrorDetail` indexed mapping isn't type-safe, `details` don't have type hints for the types. Also, I'm not sure what's the purpose of making it an indexed mapping and an array simultaneously. I'd say only one of them would make sense (preferably something closer to the first).

Instead, I'd suggest something like this:
```
class RaidenError<D> extends Error {
  details?: D;

  constructor(message?: string, details?: D) {
    super(message);
    this.details = details;
  }
}

class PfsError extends RaidenError<{ pfs: string }> {}
// now, PfsError's details is typesafe
throw new PfsError('Err1', { pfs: 'bar' }); // correct
throw new PfsError('Err2', { pfs: 123 }); // error
```
Also, I'm not sure about the purpose of the `addDetail` method, as usually error objects are mostly read-only and defined at instantiation time with all details included.

The above approach is serializable as per `ErrorCodec` (which could/should also be moved to here), although the details wouldn't be serializable. For that, you would need to make `RaidenError` and/or derivatives codec-based, or at least the details and have some kind of type mapping somewhere to find the right class and use it to decode/deserialize/instantiate the error object.

Reply:
> Suggestion: this ErrorDetail indexed mapping isn't type-safe

I don't see any reason for a need to be type safe, can you come up with scenarios?

> I'm not sure what's the purpose of making it an indexed mapping and an array simultaneously

I guess that's more a taste thing, thanks for the feedback though

> Also, I'm not sure about the purpose of the addDetail method, as usually error objects are mostly read-only and defined at instantiation time with all details included.

It's there so we can prevent having multi line throw statements where too many details are getting arranged in the constructor, as some details might need some conversions before it gets thrown.

> The above approach is serializable as per ErrorCodec (which could/should also be moved to here), although the details wouldn't be serializable. For that, you would need to make RaidenError and/or derivatives codec-based, or at least the details and have some kind of type mapping somewhere to find the right class and use it to decode/deserialize/instantiate the error object.

You shouldn't put codec types into details, that's why `ErrorDetail` was `{ [key: string]: string }`. Details should always be a simple JSON object

Reply:
> I don't see any reason for a need to be type safe, can you come up with scenarios?

I saw that we already push codecs into errors/payload/actions so I guess no way around it

Reply:
This was due to #759 , but more for consistency, we don't serialize error actions yet (only Confirmable actions are serialized, which are success actions), and I'm not sure we'll need to. Anyway, we don't push codecs to payloads, the schema declaration contain codecs, so when needing to [de]serialize/validate it, we know where to look into. Right now, only for generic Error, and as I said, due to our usage, should be enough.
But these are two different things though: typed detail (what I suggested) and serializable/codec-based errors.
Thea bove isn't the only reason to have details typed in the errors though. Internally, we make a very limited use of the errors, but if the idea is to communicate it to the user, a valid usecase may be:
```ts
try {
  await raiden.findPFS(...);
} catch (e) {
  if (isError(e, PfsError)) { // or some proper typeguard based on `name` or `message`/`ErrorType`
    showDialog(`Could not contact PFS at ${e.detail.pfs}`); // <- type-checked
  } else {
    showDialog('A generic error ocurred');
  }
}
```

Reply:
Not a fan. There're too many cases where errors need different types of infos. For example, if:
* user disables PFS, and
* there is no direct route

We throw an error at some point. Having a pfs prop there makes no sense. And there many more cases where enforcing a prop makes no sense

Reply:
These cases should be different Pfs error classes, with different detail types. My example was just a toy example to show how it'd help.